Handle any exception thrown while generating source for an IngestDocument

joegallo opened the pull request:
Related to #91977

Regardless of what exception is thrown while generating source from an `IngestDocument`, we want to catch and handle the exception -- the contract for `source(..., true)` is that it can throw an `IllegalArgumentException` if there are self references, but now that we've seen other things go wrong inside the guts of that method (and given how hard they have been to diagnose!) I think a wider `catch` is justified (and remains justified even in a world where #91977 has been fixed).

joegallo committed:
Handle any exception thrown while generating source

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo committed:
Update docs/changelog/91981.yaml

joegallo committed:
Widen the wrapping exception

It seems misleading to use IllegalArgumentException for the general
case of "something, anything, went wrong here".

masseyke reviewed (approved):
LGTM. As we've discussed offline I think we ought to broaden where we catch exceptions (since there are other places that would also result in the client never being notified). But this is definitely the most urgent and lowest risk place (since we have seen an exception right here in production).

joegallo committed:
Use two catch blocks

The first catch block converts to an IllegalArgumentException which
maps to a 400. The second block converts to a RuntimeException which
maps to a 500.

For bwc reasons, we don't want to change the first block, but at the
same time it's not right to return a 400 for the second situation --
we don't know what happened and can't say for sure it's a problem with
the request.

joegallo committed:
Reword this comment slightly

joegallo commented:
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

masseyke reviewed (approved):
LGTM
